### PR TITLE
feat(vic3): validate create_building

### LINKED ISSUE #125: Check whether the items of `activate_production_methods` inside `create_building` is valid for the given building type

### PR DESCRIPTION
**Pull request by amtep**
Also fixes #125

### CODE REVIEW

File: src/vic3/data/buildings.rs [hunk 1]
```diff
@@ -24,6 +25,29 @@ impl BuildingType {
...
+        if let Some(groups) = block.get_field_list("production_method_groups") {
+            for group in groups {
+                if let Some((_, block, group_item)) = data
+                    .get_item::<ProductionMethodGroup>(Item::ProductionMethodGroup, group.as_str())
+                {
+                    if group_item.contains_production_method(pm, block, data) {
+                        return;
+                    }
+                }
+            }
```
Comment (my4ng) [inline src/vic3/data/buildings.rs:46]: This looks like only the first group is checked if it does contain production method. Is this the desired behaviour?

File: src/vic3/data/production_methods.rs [hunk 2]
```diff
@@ -99,6 +99,19 @@ impl ProductionMethodGroup {
...
+    pub fn contains_production_method(
+        &self,
+        contains_pm: &Token,
+        block: &Block,
+        _data: &Everything,
+    ) -> bool {
+        if let Some(pms) = block.get_field_list("production_methods") {
+            return pms.iter().any(|pm| pm == contains_pm);
+        }
+        false
```
Comment (my4ng) [inline src/vic3/data/production_methods.rs:113]: can be replaced by `is_some_and` but not a big deal

File: src/vic3/effect_validation.rs [hunk 3]
```diff
@@ -243,6 +244,57 @@ pub fn validate_country_value(
...
+        if let Some(building) = building {
+            if let Some((_, block, building_item)) =
+                data.get_item::<BuildingType>(Item::BuildingType, building.as_str())
+            {
+                building_item.validate_production_method(token, building, block, data);
+            }
+        }
+    });
+    vd.field_bool("subsidized");
+    vd.field_numeric_range("reserves", 0.0..=1.0);
```
Comment (my4ng) [inline src/vic3/effect_validation.rs:269]: There are no `reserves=0.*` in vanilla, but there is one with `=2`. Maybe this is in absolute unit?

File: src/vic3/effect_validation.rs [hunk 4]
```diff
@@ -243,6 +244,57 @@ pub fn validate_country_value(
...
+    vd.field_validated_value("level", |_, mut vd| {
+        vd.maybe_is("arable_land");
```
Comment (my4ng) [inline src/vic3/effect_validation.rs:271]: didn't find a reference to this usage in vanilla

File: src/vic3/effect_validation.rs [hunk 5]
```diff
@@ -243,6 +244,57 @@ pub fn validate_country_value(
...
+        vd.integer();
+    });
+    vd.field_validated_block("add_ownership", |block, data| {
+        let mut vd = Validator::new(block, data);
+        vd.multi_field_validated_block("country", |block, data| {
+            let mut vd = Validator::new(block, data);
+            vd.req_field("country");
+            vd.field_target("country", sc, Scopes::Country);
+            vd.req_field("levels");
+            vd.field_script_value("levels", sc);
```
Comment (my4ng) [inline src/vic3/effect_validation.rs:281]: might be hardcoded `field_integer`

File: src/vic3/effect_validation.rs [hunk 6]
```diff
@@ -243,6 +244,57 @@ pub fn validate_country_value(
...
+        });
+        // Docs say "country" for both, but vanilla uses "building".
+        vd.multi_field_validated_block("building", |block, data| {
+            let mut vd = Validator::new(block, data);
+            vd.req_field("country");
+            vd.field_target("country", sc, Scopes::Country);
+            vd.req_field("levels");
+            vd.field_script_value("levels", sc);
```
Comment (my4ng) [inline src/vic3/effect_validation.rs:289]: same

**Review by my4ng — CHANGES_REQUESTED** [review]

File: src/vic3/data/buildings.rs (hunk 1 above)
Comment (amtep) [reply src/vic3/data/buildings.rs:46]: Hmm I don't see it. It's supposed to exit if any of the groups contain that production method.

File: src/vic3/effect_validation.rs (hunk 3 above)
Comment (amtep) [reply src/vic3/effect_validation.rs:269]: I got this from the effect documentation. I don't know why there's a reserves=2 or what it's supposed to mean, i figured it was a bug in vanilla.

File: src/vic3/effect_validation.rs (hunk 4 above)
Comment (amtep) [reply src/vic3/effect_validation.rs:271]: It was also in the effect documentation :)

File: src/vic3/effect_validation.rs (hunk 5 above)
Comment (amtep) [reply src/vic3/effect_validation.rs:281]: Yeah I don't know so I went with the less restrictive one. I guess I could test it

Comment (dragon-archer) [reply src/vic3/effect_validation.rs:281]: I've tested this in my mod, seems that both `level` in `create_building` and `levels` in `add_ownership` inside `create_building` accept only `field_integer`, while `script value`, `value scope` and `variable` will all generate an error.

P.S. Actually this is very annoying, so I even create a `scripted effect` to simulate `create_building` while accepting any kinds of `level` by enuming the `level` and expand it to `create_building` call with `field_integer`

Comment (amtep) [reply src/vic3/effect_validation.rs:281]: Many thanks for testing it :)

File: src/vic3/data/production_methods.rs (hunk 2 above)
Comment (amtep) [reply src/vic3/data/production_methods.rs:113]: I think that would actually make the code harder to read, with its nested closures

**Pull request merged**